### DISCUSSION TITLE
solv_ivp integrator in variational dynamics

### DISCUSSION THREAD

**Discussion by gaox-qd** [tencent-quantum-lab/tensorcircuit#32] [General]
Hi,
Naively, I replaced the propagator ```update``` function in ```variational_dynamics.ipynb```  directly with ```solv_ivp``` in ```scipy```, but I came to the problems:  ```TypeError: Expected complex128, but got Tensor("h_t0:0", shape=(), dtype=int32) of type 'Tensor'.```  Could you suggest how to resolve it. Thanks.
```python
@tc.backend.jit
def f_vqs(t, theta):
    h_t = 

    h_t = tc.quantum.PauliStringSum2Dense(
            tf.constant(h, dtype=tf.complex128), tf.constant(w, dtype=tf.complex128)
             ) 

    lhs = lhs_matrix(theta,psi0)
    rhs = rhs_vector(theta,psi0, h_t)
    #protection
    eps = 1e-4
    lhs += eps * tc.backend.eye(l * ltheta, dtype=lhs.dtype)
    return  tc.backend.solve(lhs, rhs, assume_a="sym")
result_solve_ode= solve_ivp(f_vqs,t_span,theta0,method='RK45')
```

**Comment by refraction-ray** [comment]
I think the origin of the problem may be the mix use of tf.Tensor and np.array, you can try to refactor the code as the follows and see whether it works (always pay attention to the type (tensor/array) and dtype(real/complex) of each variable).

```python
@tc.backend.jit
def f(theta, h_t):
    lhs = lhs_matrix(theta,psi0)
    rhs = rhs_vector(theta,psi0, h_t)
    eps = 1e-4
    lhs += eps * tc.backend.eye(l * ltheta, dtype=lhs.dtype)
    return  tc.backend.solve(lhs, rhs, assume_a="sym")

def f_vqs(t, theta):
    h_t = 
    theta = tc.backend.convert_to_tensor(theta)
    r = f(theta, h_t)
    return tc.backend.numpy(r)

result_solve_ode= solve_ivp(f_vqs,t_span,theta0,method='RK45')

```

**Reply by gaox-qd** [reply]
Many thanks!
Indeed I make mistake on mix using of array and tensor.
With your sugest code, it works for me now!

**Comment by refraction-ray** [comment]
And for anyone who is interested, I have added a numpy interface, so the code for ``f_vqs`` above can be simplified as 

```python
def f(theta, h_t):
    pass

f_np = tc.interfaces.numpy_interface(f)

def f_vqs(t, theta):
    h_t = 
    return f_np(theta, h_t)

```
